File: Week02/week02/二叉树前序遍历/preorderTraversal3.py
from sklearn.tree import DecisionTreeClassifier
class Solution(object):
	def preorderTraversal3(self,root):
		'''
		:type root:Treenode
		:rtype:List[int]
		'''
		node,output=root,[]
		while node:
			if not node.left:
				output.append(node.val)
				node=node.right
			else:
				predecessor=node.left
				while predecessor.right and predecessor.right is not node:
					predecessor=predecessor.right
				if not predecessor.right:
					output.append(node.val)
					predecessor.right=node
					node=node.left
				else:
					predecessor.right=None
					node=node.right
		return output

File: Week06/Week02/week02/二叉树前序遍历/preorderTraversal3.java
package yiyi;

import java.util.LinkedList;
import java.util.List;


class TreeNode{
	public void Node() {}
	TreeNode val;
	public Object left;
	public TreeNode right;
	

	public LinkedList<TreeNode> preorderTraversal3 (TreeNode root){
		LinkedList<TreeNode> output=new LinkedList<>();
		Object node=root;
		while(node!=null) {
			if(node.left==null) {
				output.add(node.val);
				node=node.right;
			}else {
				TreeNode predecessor=node.left;
				while((predecessor.right!=null)&&(predecessor.right!=node)) {
					predecessor=predecessor.right;
				}
				if(predecessor.right==null) {
					output.add(node.val);
					predecessor.right=node;
					node=node.left;
				}else {
					predecessor.right=null;
					node=node.right;
				}
			}
		}
		return output;
	}
}